src/Microsoft.AspNet.Mvc.Core/ModelBinding/Validation/DefaultObjectValidator.cs

```diff
@@ -312,6 +270,52 @@ private bool IsTypeExcludedFromValidation(IList<IExcludeTypeValidationFilter> fi
             return filters.Any(filter => filter.IsTypeExcluded(type));
         }
 
+        private void ExpandValidationNodeAndGetValidators(
+            ValidationContext context, ModelExplorer modelExplorer)
+        {
+            var validationNode = context.ValidationNode;
+            if (validationNode.ChildNodes.Count != 0 ||
+                !validationNode.ValidateAllProperties ||
+                validationNode.Model == null)
+            {
+                return;
+            }
+
+            if (!modelExplorer.Metadata.IsCollectionType)
+            {
+                foreach (var property in validationNode.ModelMetadata.Properties)
+                {
+                    var propertyExplorer = modelExplorer.GetExplorerForProperty(property.PropertyName);
+                    var propertyBindingName = property.BinderModelName ?? property.PropertyName;
+                    var childKey = ModelNames.CreatePropertyModelName(validationNode.Key, propertyBindingName);
+                    var childNode = new ModelValidationNode(childKey, property, propertyExplorer.Model)
+                    {
+                        ValidateAllProperties = true
+                    };
+                    validationNode.ChildNodes.Add(childNode);
+                }
+            }
+            else
+            {
+                var enumerableModel = (IEnumerable)modelExplorer.Model;
+                var elementType = GetElementType(enumerableModel.GetType());
+                var elementMetadata = _modelMetadataProvider.GetMetadataForType(elementType);
+                var index = 0;
```

Comment:
Add a comment that `index` would be incorrect in many scenarios but those scenarios are covered by explicit `ModelValidationNode` creation in model binding.